src/clients/c++/concurrency_manager.h

```diff
@@ -0,0 +1,188 @@
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+#include "src/clients/c++/load_manager.h"
+#include "src/clients/c++/model_context.h"
+#include "src/clients/c++/perf_utils.h"
+
+#include <condition_variable>
+#include <thread>
+
+
+namespace perfclient {
+//==============================================================================
+/// ConcurrencyManager is a helper class to send inference requests to inference
+/// server consistently, based on the specified setting, so that the perf_client
+/// can measure performance under different concurrency.
+///
+/// An instance of concurrency manager will be created at the beginning of the
+/// perf client and it will be used to simulate different load level in respect
+/// to number of concurrent infer requests and to collect per-request statistic.
+///
+/// Detail:
+/// Concurrency Manager will maintain the number of concurrent requests by
+/// spawning worker threads that keep sending randomly generated requests to the
+/// server. The worker threads will record the start time and end
+/// time of each request into a shared vector.
+///
+class ConcurrencyManager : public LoadManager {
+ public:
+  ~ConcurrencyManager();
+
+  /// Create a concurrency manager that is responsible to maintain specified
+  /// load on inference server.
+  /// \param batch_size The batch size used for each request.
+  /// \param max_threads The maximum number of working threads to be spawned.
+  /// \param sequence_length The base length of each sequence.
+  /// \param zero_input Whether to fill the input tensors with zero.
+  /// \param factory The ContextFactory object used to create InferContext.
+  /// \param manger Returns a new ConcurrencyManager object.
+  /// \return Error object indicating success or failure.
+  static nic::Error Create(
+      const int32_t batch_size, const size_t max_threads,
+      const size_t sequence_length, const bool zero_input,
+      const std::string& data_directory,
+      const std::shared_ptr<ContextFactory>& factory,
+      std::unique_ptr<LoadManager>* manager);
+
+  /// Implementation of ChangeConccurencyLevel to maintain the specified
+  /// concurency.
+  /// \parm concurent_request_count The number of concurrent requests to be
+  /// maintained.
+  /// \return Error object indicating success or failure.
```

Comment:
Since these functions have been documented in `LoadManager`, you can just refer where to find it with `// @see LoadManager.ChangeConcurrencyLevel`

Reply:
Done